File: Projects/BLUELINE/staging-skeleton/test_staging_skeleton.js
// Cross-language parity: the JS mirror reproduces the Python-generated golden fixtures EXACTLY.
// This is the real "both speak one skeleton" guarantee. Run:  node test_staging_skeleton.js
const fs = require("fs"), path = require("path");
const S = require("./staging_skeleton.js");
const DOC = JSON.parse(fs.readFileSync(path.join(__dirname, "staging-skeleton.fixtures.json"), "utf8"));

function close(a, b, tol = 1e-9) {
  if (typeof a === "boolean" || typeof b === "boolean") return a === b;
  if (typeof a === "number" && typeof b === "number") return Math.abs(a - b) <= tol;
  if (Array.isArray(a) && Array.isArray(b)) return a.length === b.length && a.every((x, i) => close(x, b[i], tol));
  if (a && b && typeof a === "object" && typeof b === "object") {
    const ka = Object.keys(a), kb = Object.keys(b);
    return ka.length === kb.length && ka.every(k => k in b && close(a[k], b[k], tol));
  }
  return a === b;
}

let fails = 0;
for (const c of DOC.cases) {
  const kp = c.keypoints;
  const checks = [
    ["frame_authored", S.stagingFrame(kp, c.authored_facing), c.frame_authored],
    ["frame_estimated", S.stagingFrame(kp, null), c.frame_estimated],
    ["facing_estimate", S.facingFromKeypoints(kp), c.facing_estimate],
  ];
  for (const [label, got, exp] of checks) {
    if (!close(got, exp)) { fails++; console.log(`FAIL ${c.name} ${label}\n  got ${JSON.stringify(got)}\n  exp ${JSON.stringify(exp)}`); }
  }
}
console.log(`JS ${fails === 0 ? "PASS" : "FAIL"} — ${DOC.cases.length} cases, ${fails} mismatch(es)`);
process.exit(fails ? 1 : 0);

File: Projects/BLUELINE/staging-skeleton/test_staging_skeleton.py
"""Regression guard: the Python reference reproduces staging-skeleton.fixtures.json.
Run:  python test_staging_skeleton.py"""
import json, os, sys
import staging_skeleton as S

HERE = os.path.dirname(os.path.abspath(__file__))
DOC = json.load(open(os.path.join(HERE, "staging-skeleton.fixtures.json")))

def close(a, b, tol=1e-9):
    if isinstance(a, bool) or isinstance(b, bool): return a == b
    if isinstance(a, (int, float)) and isinstance(b, (int, float)): return abs(a-b) <= tol
    if isinstance(a, list) and isinstance(b, list):
        return len(a) == len(b) and all(close(x, y, tol) for x, y in zip(a, b))
    if isinstance(a, dict) and isinstance(b, dict):
        return a.keys() == b.keys() and all(close(a[k], b[k], tol) for k in a)
    return a == b

fails = 0
for c in DOC["cases"]:
    kp = c["keypoints"]
    got_auth = S.staging_frame(kp, c["authored_facing"])
    got_est  = S.staging_frame(kp, None)
    got_fac  = S.facing_from_keypoints(kp)
    for label, got, exp in (("frame_authored", got_auth, c["frame_authored"]),
                             ("frame_estimated", got_est, c["frame_estimated"])):
        if not close(got, exp):
            fails += 1; print(f"FAIL {c['name']} {label}\n  got {got}\n  exp {exp}")
    if not close(got_fac, c["facing_estimate"]):
        fails += 1; print(f"FAIL {c['name']} facing_estimate got {got_fac} exp {c['facing_estimate']}")

print(f"PY {'PASS' if fails==0 else 'FAIL'} — {len(DOC['cases'])} cases, {fails} mismatch(es)")
sys.exit(1 if fails else 0)
